src/libraries/System.Private.CoreLib/src/System/Threading/Tasks/Sources/ManualResetValueTaskSourceCore.cs

```diff
@@ -207,51 +207,104 @@ private void SignalCompletion()
             }
             _completed = true;
 
-            if (_continuation != null || Interlocked.CompareExchange(ref _continuation, ManualResetValueTaskSourceCoreShared.s_sentinel, null) != null)
+            if (_continuation is null && Interlocked.CompareExchange(ref _continuation, ManualResetValueTaskSourceCoreShared.s_sentinel, null) is null)
             {
-                if (_executionContext != null)
+                return;
+            }
+
+            if (_executionContext is null)
+            {
+                if (_capturedContext is null)
+                {
+                    if (RunContinuationsAsynchronously)
+                    {
+                        ThreadPool.UnsafeQueueUserWorkItem(_continuation, _continuationState, preferLocal: true);
+                    }
+                    else
+                    {
+                        _continuation(_continuationState);
+                    }
+
+                    return;
+                }
+
+                InvokeSchedulerContinuation();
+                return;
+            }
+
+            ExecutionContext? currentContext = ExecutionContext.Capture();
+            // Restore the captured ExecutionContext before executing anything.
+            ExecutionContext.Restore(_executionContext);
+
+            if (_capturedContext is null)
+            {
+                if (RunContinuationsAsynchronously)
                 {
-                    ExecutionContext.RunInternal(
-                        _executionContext,
-                        (ref ManualResetValueTaskSourceCore<TResult> s) => s.InvokeContinuation(),
-                        ref this);
+                    ThreadPool.QueueUserWorkItem(_continuation, _continuationState, preferLocal: true);
+                    // Restore the current ExecutionContext.
+                    ExecutionContext.Restore(currentContext);
                 }
                 else
                 {
-                    InvokeContinuation();
+                    // Running inline may throw; capture the edi if it does as we changed the ExecutionContext,
+                    // so need to restore it back before propagating the throw.
+                    ExceptionDispatchInfo? edi = InvokeInlineContinuation();
```

Comment:
If I read it correctly, before the change the continuation was run directly, and after the change the sync context is captured and restored. Is that correct and is it necessary? I don't know much about the usages here.

Reply:
The if/else was switched; so this used `ExecutionContext.Run` before, which also captures the exception in an edi; restores the context, then rethrows it